File: internal/config/workflow.go
package config

type Workflow struct {
	Name string `yaml:"name" validate:"required"`

	EventType EventType `yaml:"event_type"   validate:"required,dive,required"`

	Middlewares Middlewares `yaml:"middlewares"   validate:"required"`

	Inputs  []Input  `yaml:"inputs"   validate:"required,dive,required"`
	Outputs []Output `yaml:"outputs"   validate:"required,dive,required"`
}

type EventType struct {
	Format string  `yaml:"format"   validate:"required"`
	Fields []Field `yaml:"fields"   validate:"required,dive,required"`
}

type Field struct {
	Name     string `yaml:"name"   validate:"required"`
	DataType string `yaml:"data_type"   validate:"required"`
	Required bool   `yaml:"required"`
}

type Input struct {
	Name    string  `yaml:"name"   validate:"required"`
	Source  string  `yaml:"source"   validate:"required"`
	Decoder Decoder `yaml:"decoder"   validate:"required,dive,required"`
	Params  any     `yaml:"params"   validate:"required"`
}

type Decoder struct {
	Format  string   `yaml:"format"   validate:"required"`
	Mappers []Mapper `yaml:"mappers"   validate:"required,dive,required"`
}

type Mapper struct {
	Source string `yaml:"source"   validate:"required"`
	Target string `yaml:"target"   validate:"required"`
}

type Output struct {
	Name        string     `yaml:"name"   validate:"required"`
	Destination string     `yaml:"destination"   validate:"required"`
	Params      any        `yaml:"params"   validate:"required"`
	Condition   *Condition `yaml:"condition,omitempty" `
}

type Condition struct {
	Field      string      `yaml:"field,omitempty"`
	Operator   string      `yaml:"operator"`
	Value      any         `yaml:"value,omitempty"`
	Conditions []Condition `yaml:"conditions,omitempty"`
}

type Middlewares struct {
	CORS []CORS `yaml:"cors"`
	Auth []Auth `yaml:"auth"`
}

type CORS struct {
	Name             string   `yaml:"name"`
	AllowedOrigins   []string `yaml:"allowed_origins"`
	AllowedMethods   []string `yaml:"allowed_methods"`
	AllowedHeaders   []string `yaml:"allowed_headers"`
	AllowCredentials bool     `yaml:"allow_credentials"`
	MaxAge           int      `yaml:"max_age"`
}

type Auth struct {
	Name   string `yaml:"name"`
	Type   string `yaml:"type"`
	Params any    `yaml:"params"`
}

File: doc/src/components/Editor/workflow.js
export const workflow = `name: incoming-orders

event_type:
  format: json
  fields:
    - name: customerId
      data_type: string
      required: true
    - name: orderNumber
      data_type: int
      required: true
    - name: orderStatus
      data_type: string
      required: true
    - name: nonRequiredField
      data_type: string

middlewares:
  cors:
    default:
      allowed_origins:
        - http://localhost:3000
        - http://localhost:4000
      allowed_methods:
        - POST
      allowed_headers:
        - Content-Type
        - Authorization
      allow_credentials: true
      max_age: 3600

  auth:
    default:
      type: token
      params:
        token: 2dc7929e5b589cb7861bcae19e13ad96
        query_param: token

inputs:
  - name: webhook-orders
    source: webhook
    params:
      endpoint: /webhook/orders
      auth: default
    decoder:
      format: json
      mappers:
        - source: o_customer_id
          target: customerId
        - source: o_number
          target: orderNumber
        - source: o_status
          target: orderStatus

outputs:
  - name: sse-orders-completed
    destination: sse
    params:
      endpoint: /sse/orders/completed
      auth: default
      cors: default
    encoder:
      format: json
    condition:
      operator: or
      conditions:
        - field: orderStatus
          operator: eq
          value: "completed"
        - field: orderStatus
          operator: eq
          value: "pending"

  - name: smtp-orders-failed
    destination: smtp
    params:
      from: hi@valensto.com
      to: v.e.brochard@gmail.com
      subject: Order failed
      base_url: http://localhost:4000
      host: smtp.gmail.com
      port: "587"
      username: hi@valensto.com
      password: bbfrhiypmitrixyw
      enable_starttls: true
    encoder:
      format: html
    condition:
      field: orderStatus
      operator: eq
      value: "failed"`